runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,517 @@
+        txBeginAndApply(table); // Apply pending writes (if any) & begin corfu txn if thread hasn't begun one
+        table.getMetrics().numGetByIndex.incrementAndGet();
+        return table.getByIndex(indexName, indexKey);
+    }
+
+    /**
+     * Gets the count of records in the table at a particular timestamp.
+     *
+     * @param table - the table whose count is requested.
+     * @return Count of records.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    int count(@Nonnull final Table<K, V, M> table) {
+        txBeginAndApply(table); // Apply pending writes (if any) & begin corfu txn if thread hasn't begun one
+        table.getMetrics().numCounts.incrementAndGet();
+        return table.count();
+    }
+
+    /**
+     * Gets all the keys of a table.
+     *
+     * @param table - the table whose keys are requested.
+     * @return keyset of the table
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    Set<K> keySet(@Nonnull final Table<K, V, M> table) {
+        txBeginAndApply(table); // Apply pending writes (if any) & begin corfu txn if thread hasn't begun one
+        table.getMetrics().numKeySets.incrementAndGet();
+        return table.keySet();
+    }
+
+    /**
+     * Scan and filter by entry.
+     *
+     * @param table Table< K, V, M > object on which the scan must be done.
+     * @param entryPredicate Predicate to filter the entries.
+     * @return Collection of filtered entries.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    List<CorfuStoreEntry<K, V, M>> executeQuery(@Nonnull final Table<K, V, M> table,
+                                                @Nonnull final Predicate<CorfuStoreEntry<K, V, M>> entryPredicate) {
+        txBeginAndApply(table); // Apply pending writes (if any) & begin corfu txn if thread hasn't begun one
+        table.getMetrics().numScans.incrementAndGet();
+        return table.scanAndFilterByEntry(entryPredicate);
+    }
+
+    /**
+     * Start the actual corfu transaction if started by a read operation,
+     * Cases to consider:
+     *   Bit2      Bit1         Bit0
+     *  isRead   hasTxnBegun   isEmptyWriteBuffer
+     *  ======= ===========   ================
+     * 0 commit noTxnStarted  buffered writes => start txn, apply writes, writeOnlyTxn
+     * 1 commit noTxnStarted  no writes       => start txn, writeOnlyTxn
+     * 2 commit txnStarted    buffered writes => apply writes, writeAndReadTxn
+     * 3 commit txnStarted    no writes       => do nothing!
+     * 4 readApi noTxnStarted buffered writes => start txn, apply writes, writeAndReadTxn
+     * 5 readApi noTxnStarted no writes       => start txn, readOnlyTxn
+     * 6 readApi txnStarted   buffered writes => apply writes, writeAndReadTxn
+     * 7 readApi txnStarted   no writes       => readOnlyTxn iff not writeAndReadTxn
+     *
+     * @param tableBeingRead - am i being called from read api
```

Comment:
can you change the api as per the comment or vice versa?

Reply:
done. fix up the comment.